### DISCUSSION TITLE
Using this.$on on constructor got error

### DISCUSSION THREAD

**Discussion by evmartinelli** [prisma/prisma#23831] [Q&A] [answered]
### Question

<img width="766" alt="image" src="https://github.com/prisma/prisma/assets/2855688/d2790568-88fe-4e13-bfbf-41f45d13b6fa">

What am I doing wrong?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
HI @evmartinelli 👋 

Thank you for raising this question.

The issue is likely due to TypeScript not being able to infer the correct type for the `$on` method when you're extending the PrismaClient class. You need to pass the generics yourself when extending the PrismaClient class. 

```ts
import { Injectable } from '@nestjs/common';
import { PrismaClient, Prisma } from '@prisma/client';

@Injectable()
export class PrismaService extends PrismaClient<Prisma.PrismaClientOptions, "query"> {
  constructor() {
    super({ log: [{ level: 'query', emit: 'event' }]});
    this.$on('query', (e) => {
      console.log('Query executed: ' + e.query);
      console.log('Duration: ' + e.duration + 'ms');
    });
  }
}
```
The generics `<Prisma.PrismaClientOptions, "query">` are passed to PrismaClient when extending it. This helps TypeScript understand the type of the first argument of `$on`, which is only enabled when the user passes some option to the client.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by evmartinelli** [reply]
Thank you so much! Now it works and great explanation.
